### PR TITLE
New: the `prefer-spread` rule (fixes #2946)

### LINKED ISSUE #2946: new rule: prefer-spread
In ES6 there is the wonderful spread operator `...` that makes it easy to apply an array to a function.

In JS you may have previously typed `console.log.apply(console, [1,2,3])` or `foo.apply(foo, x)`.

With spread you can instead say `console.log(...[1,2,3])` or `foo(...x)`.

In my opinion it's a much nicer syntax and generally should be preferred over the other when in an ES6 environment.

**Caveats**

You can't easily imitate this with spread:
`foo.apply(bar, x)`

So the rule would limit itself to occasions where the `context` is the same as the function being applied to. For that case it was suggested that `Reflect.apply` should be used instead (See: #2939)

### PR DESCRIPTION
**Pull request by mysticatea**
:)

Note: To compare `this` bindings, I use just text comparison, so this rule warns `obj[++i].foo.apply(obj[++i], args)`.  But I avoided to be complex.

### CODE REVIEW

File: docs/rules/prefer-spread.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Suggest using of the spread operator instead of `.apply()` callings. (prefer-spread)
+
+Up to now, we had been using `Function.prototype.apply()` callings for the variadic arguments.
```
Comment (ilyavolodin) [inline docs/rules/prefer-spread.md]: I think term `varidic arguments` only applies to `...` operator in C++. The more common term is `varidic function`. Please correct me if I'm wrong.

File: docs/rules/prefer-spread.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+Since ES2015, we can use the spread operator for the variadic arguments.
+
+```js
+var args = [1, 2, 3, 4];
+Math.max(...args);
+```
+
+## Rule Details
+
+This rule is aimed to flag `Function.prototype.apply()` callings can be replaced to the spread operator.
```
Comment (ilyavolodin) [inline docs/rules/prefer-spread.md]: This rule is aimed to flag usage of `Function.prototype.apply()` that can be replaced with the spread operator.

File: docs/rules/prefer-spread.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+obj.foo.apply(null, args); // the `this` binding is different.
+```
+
+```js
+obj.foo.apply(otherObj, args); // the `this` binding is different.
+```
+
+## When Not to Use It
+
+If you don't want to be notified about `Function.prototype.apply()` callings, you can safely disable this rule.
```
Comment (ilyavolodin) [inline docs/rules/prefer-spread.md]: You should also mention that this rule should not be used in ES3/5 environments.

File: docs/rules/README.md [hunk 4]
```diff
@@ -204,6 +204,7 @@ These rules are only relevant to ES6 environments.
 * [no-var](no-var.md) - require `let` or `const` instead of `var`
 * [object-shorthand](object-shorthand.md) - require method and property shorthand syntax for object literals
 * [prefer-const](prefer-const.md) - suggest using of `const` declaration for variables that are never modified after declared
+* [prefer-spread](prefer-spread.md) - suggest using of the spread operator instead of `.apply()` callings.
```
Comment (ilyavolodin) [inline docs/rules/README.md]: `calling` at the end is not necessary.

**Comment by ilyavolodin** [comment]
LGTM otherwise.

**Comment by mysticatea** [comment]
Thank you a lot for the review!

I fixed, the diff is here: https://github.com/mysticatea/eslint/commit/9980f79fb7a246dbf3eab3a94bfc3c36ab0d1731

I learned today, the word of `variadic function`. :)

**Comment by ilyavolodin** [comment]
Thanks for the quick turn around.

**Pull request merged**